lib/tapioca/dsl/compilers/json_api_client_resource.rb

```diff
@@ -0,0 +1,208 @@
+          def gather_constants
+            all_modules.select do |c|
+              name_of(c) && c < ::JsonApiClient::Resource
+            end
+          end
+        end
+
+        private
+
+        sig { returns(T.nilable(::JsonApiClient::Schema)) }
+        def resource_schema
+          schema = constant.schema
+
+          # empty? does not exist on JsonApiClient::Schema
+          schema if schema.size > 0 # rubocop:disable Style/ZeroLengthPredicate
+        end
+
+        sig do
+          params(
+            mod: RBI::Scope,
+            property: ::JsonApiClient::Schema::Property,
+          ).void
+        end
+        def generate_methods_for_property(mod, property)
+          type = type_for(property)
+
+          name = property.name.to_s
+
+          mod.create_method(name, return_type: type)
+          mod.create_method("#{name}=", parameters: [create_param(name, type: type)], return_type: type)
+        end
+
+        sig { params(property: ::JsonApiClient::Schema::Property).returns(String) }
+        def type_for(property)
+          type = ::JsonApiClient::Schema::TypeFactory.type_for(property.type)
+          return "T.untyped" if type.nil?
+
+          sorbet_type = if type.respond_to?(:sorbet_type)
+            type.sorbet_type
+          elsif type == ::JsonApiClient::Schema::Types::Integer
+            "::Integer"
+          elsif type == ::JsonApiClient::Schema::Types::String
+            "::String"
+          elsif type == ::JsonApiClient::Schema::Types::Float
+            "::Float"
+          elsif type == ::JsonApiClient::Schema::Types::Time
+            "::Time"
+          elsif type == ::JsonApiClient::Schema::Types::Decimal
+            "::BigDecimal"
+          elsif type == ::JsonApiClient::Schema::Types::Boolean
+            "T::Boolean"
+          else
+            "T.untyped"
+          end
+
+          if property.default.nil?
+            as_nilable_type(sorbet_type)
+          else
+            sorbet_type
+          end
+        end
+
+        sig do
+          params(
+            mod: RBI::Scope,
+            association: JsonApiClient::Associations::BaseAssociation,
+          ).void
+        end
+        def generate_methods_for_association(mod, association)
+          # If the association is broken, it will raise a NameError when trying to access the association_class
+          klass = association.association_class
+
+          name, type = case association
+          when ::JsonApiClient::Associations::BelongsTo::Association
+            # id must be a string: # https://jsonapi.org/format/#document-resource-object-identification
+            [association.param.to_s, "T.nilable(::String)"]
+          when ::JsonApiClient::Associations::HasOne::Association
+            [association.attr_name.to_s, "T.nilable(#{klass})"]
+          when ::JsonApiClient::Associations::HasMany::Association
+            [association.attr_name.to_s, "T.nilable(T::Array[#{klass}])"]
```

Comment:
Unfortunately, all of those things can be nullable when an object is created without being loaded via the API